### DISCUSSION TITLE
Adding cmake generated headers to includedirs

### DISCUSSION THREAD

**Discussion by ignatz** [xmake-io/xmake#4952] [Q&A]
Hi makers,

Let me start by saying, I'm probably just holding it wrong but even after looking at dozens of xmake.lua files I haven't found the right incantation yet.

I'm trying to add the dawn webgpu implementation as a dependency. Dawn's cmake build process builds headers under "<builddir>/gen/include". I've tried various versions of what I've found, i.e. 

```lua
package("webgpu")
    add_deps("cmake")
    set_sourcedir(path.join(os.scriptdir(), "vendor/dawn"))

    add_includedirs("$(buildir)/gen/include")

    on_load(function (package)
	package:add("includedirs", path.join(package:buildir(), "gen", "include"))
    end)

    on_install(function (package)
        local configs = {}

        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))

        import("package.tools.cmake").install(package, configs)
    end)

    on_test(function (package)
	print(package)
        assert(package:has_cxxfuncs("wgpuCreateInstance", {
		includes = "dawn/webgpu.h"
	}))
    end)
package_end()
```

printing the package definition contains:

```
      includedirs = {
        "$(buildir)/gen/include",
        "/<redacted>/build/.packages/w/webgpu/latest/cache/build_4e0143c9/gen/include"
      },
```

So the `$(buildir)` approach didn't expand (not sure why) but the on_load approach seemed to have worked. Yet, the `on_test_ cxxfuncs test fails with:

```
> checking for c++ snippet(has_cxxfuncs)
error: /tmp/.xmake1000/240408/_678DB87C6E4D4C789BF1B9B9A88A30BA.cpp:2:10: fatal error: dawn/webgpu.h: No such file or directory
    2 | #include <dawn/webgpu.h>
      |          ^~~~~
```

I didn't find a way to see the actual compiler invocation but I assume it doesn't contain the includedirs listed above, otherwise I'd expect it to work.

Just double checking the above include path:

```
$ ls /<redacted>/build/.packages/w/webgpu/latest/cache/build_4e0143c9/gen/include/dawn/                                                       ~/
dawn_proc_table.h  webgpu_cpp.h  webgpu_cpp_print.h  webgpu.h
```
it points to the correct generated artifacts.

FYI, I'm using xmake `xmake v2.8.7+20240227` from the debian repos.

Any thoughts?

**Comment by waruqi** [comment]
If webgpu has a cmake install script, then calling cmake.install() in on_install will automatically add its headers to the package installation directory. Please check cmakelists.txt and the package installation directory to make sure the installed headers exist.

These are not necessary.
> 
    add_includedirs("$(buildir)/gen/include")

    on_load(function (package)
	package:add("includedirs", path.join(package:buildir(), "gen", "include"))
    end)

**Comment by waruqi** [comment]
If you install failed, you can check the failed installation directory `installdir.failed`. like this.

```
ls ~/.xmake/cache/packages/2404/z/zlib/v1.3.1/installdir.failed/include/
```

If header files is missing in include, you need copy and install them in on_install

```lua

    on_install(function (package)
        local configs = {}

        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
        table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))

        import("package.tools.cmake").install(package, configs)

        os.cp("include dir", package:installdir("include")) -- TODO
    end)
```

**Comment by ignatz** [comment]
@waruqi  you're a machine :clap:. It's a treat to see you run the project, answer everyone's questions... Thank you.

You're answers helped me understand the concepts better (e.g. I was biased from other build systems to add more include paths rather than copying files into a default include path) and pointed in the right direction. You're right that dawn doesn't install the headers by default. There's extra flags for that. After fixing that I still had issues with the cmake build itself. It's almost like build wasn't deterministic because a dependency edge was missing. I guess dawn cmake is an afterthought and they use gn (or else) for Chrome :woman_shrugging: . I tried the gn integration and ran into other issues.

Since I'm not to hung up on any particular webgpu implementation I also tried wgpu-native, which has a broken meson build. It builds am empty  static library archive :sigh:

None of the above mentioned issues have anything to do with xmake. I'll keep trying. I don't wanna steal anymore of your time :pray:
